src/Spanner/Database.php

```diff
@@ -0,0 +1,1313 @@
+     * }
+     * @return Snapshot
+     * @codingStandardsIgnoreEnd
+     */
+    public function snapshot(array $options = [])
+    {
+        // These are only available in single-use transactions.
+        if (isset($options['maxStaleness']) || isset($options['minReadTimestamp'])) {
+            throw new \BadMethodCallException(
+                'maxStaleness and minReadTimestamp are only available in single-use transactions.'
+            );
+        }
+
+        $transactionOptions = $this->configureSnapshotOptions($options);
+
+        $session = $this->selectSession(SessionPoolInterface::CONTEXT_READ);
+
+        try {
+            return $this->operation->snapshot($session, $transactionOptions);
+        } finally {
+            $session->setExpiration();
+        }
+    }
+
+    /**
+     * Execute Read/Write operations inside a Transaction.
+     *
+     * Using this method and providing a callable operation provides certain
+     * benefits including automatic retry when a transaction fails. In case of a
+     * failure, all transaction operations, including reads, are re-applied in a
+     * new transaction.
+     *
+     * If a transaction exceeds the maximum number of retries,
+     * {@see Google\Cloud\Core\Exception\AbortedException} will be thrown. Any other
+     * exception types will immediately bubble up and will interrupt the retry
+     * operation.
+     *
+     * Please note that once a transaction reads data, it will lock the read
+     * data, preventing other users from modifying that data. For this reason,
+     * it is important that every transaction commits or rolls back as early as
+     * possible. Do not hold transactions open longer than necessary.
+     *
+     * If you have an active transaction which was obtained from elsewhere, you
+     * can provide it to this method and gain the benefits of managed retry by
+     * setting `$options.transaction` to your {@see Google\Cloud\Spanner\Transaction}
+     * instance. Please note that in this case, it is important that ALL reads
+     * and mutations MUST be performed within the runTransaction callable.
+     *
+     * Example:
+     * ```
+     * $transaction = $database->runTransaction(function (Transaction $t) use ($username, $password) {
+     *     $rows = $t->execute('SELECT * FROM Users WHERE Name = @name and PasswordHash = @password', [
+     *         'parameters' => [
+     *             'name' => $username,
+     *             'password' => password_hash($password, PASSWORD_DEFAULT)
+     *         ]
+     *     ])->rows();
+     *     $user = $rows->current();
+     *
+     *     if ($user) {
+     *         // Do something here to grant the user access.
+     *         // Maybe set a cookie?
+     *
+     *         $user['loginCount'] = $user['loginCount'] + 1;
+     *         $t->update('Users', $user);
+     *
+     *         $t->commit();
+     *     } else {
+     *         $t->rollback();
```

Comment:
This leaves scope for users to forget to call either or call both. Why not do this automatically for them? They can throw an exception to indicate that they want the txn to be rolled back otherwise we commit.